src/NLog/Internal/PropertiesDictionary.cs

```diff
@@ -0,0 +1,642 @@
+        /// </summary>
+        /// <param name="messageProperties">Message-template-parameters</param>
+        /// <param name="eventProperties">The initially empty dictionary</param>
+        /// <returns>Message-template-parameters was inserted into dictionary without trouble (true/false)</returns>
+        private static bool InsertMessagePropertiesIntoEmptyDictionary(IList<MessageTemplateParameter> messageProperties, Dictionary<object, PropertyValue> eventProperties)
+        {
+            try
+            {
+                for (int i = 0; i < messageProperties.Count; ++i)
+                {
+                    eventProperties.Add(messageProperties[i].Name, new PropertyValue(messageProperties[i].Value, true));
+                }
+                return true; // We are done
+            }
+            catch (ArgumentException)
+            {
+                // Duplicate keys found, lets try again
+                for (int i = 0; i < messageProperties.Count; ++i)
+                {
+                    eventProperties.Remove(messageProperties[i].Name);
+                }
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Attempt to override the existing dictionary values using the message-template-parameters 
+        /// </summary>
+        /// <param name="messageProperties">Message-template-parameters</param>
+        /// <param name="eventProperties">The already filled dictionary</param>
+        /// <returns>List of unique message-template-parameters</returns>
+        private static IList<MessageTemplateParameter> CreateUniqueMessagePropertiesListSlow(IList<MessageTemplateParameter> messageProperties, Dictionary<object, PropertyValue> eventProperties)
+        {
+            List<MessageTemplateParameter> messagePropertiesUnique = null;
+            PropertyValue valueItem;
+            for (int i = 0; i < messageProperties.Count; ++i)
+            {
+                if (eventProperties.TryGetValue(messageProperties[i].Name, out valueItem))
+                {
+                    if (valueItem.MessageProperty)
+                    {
+                        if (messagePropertiesUnique == null)
+                        {
+                            messagePropertiesUnique = new List<MessageTemplateParameter>(messageProperties.Count);
+                            for (int j = 0; j < i; ++j)
+                            {
+                                messagePropertiesUnique.Add(messageProperties[j]);
+                            }
+                        }
+                        continue;   // Skip already exists
+                    }
+                }
+
+                eventProperties[messageProperties[i].Name] = new PropertyValue(messageProperties[i].Value, true);
+                if (messagePropertiesUnique != null)
+                {
+                    messagePropertiesUnique.Add(messageProperties[i]);
```

Comment:
I think we could simplify this code by yielding messageProperties (messageProperties[i] and messageProperties[y], what do you think?

Reply:
Never been a big fan of yield because of the state-machine it allocates.